## peak_caller

Read in fastq files from Illumina sequencing, check average quality of reads, trim unsatisfied reads to desired length, 
map trimmed reads back to a reference genome, and call peaks based on given mapping threshold and peak distribution.


### read in input fastq file

### check average quality of all reads

### trim left and right border of reads to generated satisfied quality range

### map trimmed reads back to reference genome

### call peak based on given threshold and distribution width



	

	
	
	
